[JENKINS-55305] Add user creation listener methods to SecurityListener

davidolorundare opened the pull request:
See [JENKINS-55305](https://issues.jenkins-ci.org/browse/JENKINS-55305) and [JENKINS-55307](https://issues.jenkins-ci.org/browse/JENKINS-55307).

- Two new methods were added to the jenkins-core `SecurityListener` class to enable support for listening to Jenkins user-account creation events. The rationale for the additions is discussed in the above ticket.
- Implementations which exercise these `SecurityListener` methods were made in the user-creation methods of the `HudsonPrivateSecurityRealm` class.

### Proposed changelog entries

* Enhancement: Update to `jenkins.security.SecurityListener` to support notification of Jenkins user-account creation events.
* Enhancement: Update to `hudson.security.HudsonPrivateSecurityRealm` adding notification(s) of user-account creation success events to its methods.

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers
@jvz 
@jeffret-b 
@daniel-beck

davidolorundare committed:
Update core to support user-creation notifications

[JENKINS-55305] Add user creation listener method

Four new methods added to the SecurityListener class
to enable support for listening to Jenkins user-account
creation events.

[JENKINS-55305] Update SecurityListener class

[JENKINS-55307] Add new private-realm unit tests

Added unit tests to exercise the updated methods
of the HudsonPrivateSecurityRealm class.

[JENKINS-55307] Update realm user-creation methods

[JENKINS-55305] Remove usage of listener method

Updated the SecurityListener and HudsonPrivateSecurityRealm
classes, removing usage of the SecurityListener failedToCreateUser()
method, as no valid Jenkins use cases could be identified for the
method at this point in time.

jvz commented on core/src/main/java/jenkins/security/SecurityListener.java:
```diff
@@ -69,13 +69,27 @@ protected void failedToAuthenticate(@Nonnull String username){}
+    /**
+     * Fired when a new user account has been created.
+     *
+     * @param username the user
```
Might want to add a `@since TODO` javadoc line. A maintainer can go and replace those with the release version.

jvz commented on core/src/main/java/jenkins/security/SecurityListener.java:
```diff
@@ -69,13 +69,27 @@ protected void failedToAuthenticate(@Nonnull String username){}
+     *
+     * @param username the user
+     */
+    protected void failedToCreateUser(@Nonnull String username){}
```
Now I know this isn't the exact pattern that exists here, but what would you think about adding a second parameter like `@CheckForNull Exception cause` so that the exception that would otherwise be thrown later can be handled here as well. That wouldn't be required for our specific use case, though I'm curious if anyone thinks that would be useful to include anyways.

jvz replied:
By the way, if we decide to add listeners that allow for an exception as well, we can likely add those later.

jvz reviewed (approved):
LGTM other than the comments I've left.

jeffret-b commented:
I'm not clear on the use for these new methods. Are they only relevant for the internal, private security realm?

jvz reviewed (changes requested):
You need to add calls to these listeners in the appropriate places in HudsonPrivateSecurityRealm as well to help demonstrate the use case for this. Plus, we need that implemented anyways in order for this listener to be useful.

jeffret-b commented:
That makes sense. This operations aren't necessarily meaningful for the other domains.

It would be good to provide the use for these new operations.

davidolorundare commented:
@jeffret-b Hi Jeff, as Matt mentioned the use and calls to these new methods are primarily in `HudsonPrivateSecurityRealm` which are implemented in a separate pull-request that I was intending to push after this current one. I sent you some more context in a PM via Gitter.

jvz reviewed (changes requested):
Updated tasks:

- [ ] Add appropriate calls to `fireUserCreated()` and `fireFailedToCreateUser()` in `HudsonPrivateSecurityRealm`
- [ ] Add tests to `HudsonPrivateSecurityRealmTest` exercising these two new methods

davidolorundare commented:
As discussed earlier, the use and calls to these new methods are implemented in `HudsonPrivateSecurityRealm` (and the unit tests that exercise them) have now been added along with this same pull request.

jvz reviewed (approved):
Looks good to me!

jvz commented:
See also [JENKINS-55307](https://issues.jenkins-ci.org/browse/JENKINS-55307) (might want to include that in your changelog in the description).

jeffret-b reviewed (approved):
Looks good.

daniel-beck commented on core/src/main/java/jenkins/security/SecurityListener.java:
```diff
@@ -69,13 +69,31 @@ protected void failedToAuthenticate(@Nonnull String username){}
+     *
+     * @param username the user
+     */
+    protected void failedToCreateUser(@Nonnull String username){}
```
This signature does not look particularly useful without a cause (and note the possibility of 0…N+1 causes).

In fact, this seems to be lacking a use case. What's the benefit of knowing that someone didn't enter an email address?

davidolorundare replied:
as discussed earlier, I'll be pulling this method back. Will reintroduce it later if we find a more valid use case for it

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare commented:
the incremental version still doesn't get generated with this updated PR.

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare commented:
spoke with Daniel about this and decided to build the Jenkins version of this PR locally (with `deploy`) , in order to get a [Jenkins snapshot](http://repo.jenkins-ci.org/snapshots) that can be referenced in the downstream POM for later tasks of the [project](https://github.com/jenkinsci/audit-log-plugin)

davidolorundare commented:
a snapshot of this Jenkins build is located here- https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/main/jenkins-core/2.158-SNAPSHOT/

batmat commented:
@davidolorundare quick note: we would generally use the timestamped version `2.158-20190107.194411-1` in such case. 
It's the only simple way we can be 100% sure a downstream PR is using the version we think to check something.

davidolorundare commented:
@batmat I tried build my code with `2.158-20190107.194411-1` as the jenkins version number but got the error:
```
[ERROR] Failed to execute goal on project audit-log: Could not resolve dependencies for project org.jenkins-ci.plugins:audit-log:hpi:1.0-SNAPSHOT: Failure to find org.jenkins-ci.main:jenkins-war:war:2.158-20190107.194411-1 in https://repo.jenkins-ci.org/public/ was cached in the local repository, resolution will not be reattempted until the update interval of repo.jenkins-ci.org has elapsed or updates are forced```
any thoughts ?

jvz commented:
@davidolorundare you need to temporarily add the snapshots repo to the `<repositories>` section in pom.xml to support downloading that snapshot.

davidolorundare commented:
@jvz are you referring to `https://repo.jenkins-ci.org/snapshots/` or `maven.jenkins-ci.org` ?

jvz commented:
> @jvz are you referring to `https://repo.jenkins-ci.org/snapshots/` or `maven.jenkins-ci.org` ?

The first one.

davidolorundare commented:
@batmat adding in a temporary repo when using the timestamped-version, was wondering how the `repository` section of the `pom.xml` would look like ?

daniel-beck commented:
@jvz https://github.com/jenkinsci/plugin-pom/blob/c720052f7b507501b9f4bf3a32011f60c54e3561/pom.xml#L829 combines both snapshots and releases, that's not the problem.

@davidolorundare Your problem is https://github.com/jenkinsci/plugin-pom/blob/c720052f7b507501b9f4bf3a32011f60c54e3561/pom.xml#L52...L55 -- snapshots have slightly different timestamps between jenkins-core and jenkins-war, so you need to set the specific properties rather than just `jenkins.version`.

davidolorundare commented:
thanks @daniel-beck

davidolorundare commented:
@batmat @daniel-beck could a `ready-for-merge` label be placed on this PR

Wadeck reviewed (approved):
Good idea!

My only concern is about the description of the event. When do we expect other security realm to trigger it ?

As a consumer, without any indication, I expect that the user is already created **and saved** which is not the case in your implementation. Depending on the usage of such event, that could be misleading.

It was a problem with the other events. The lack of precise description made some realms to implement it without any preconditions and so, implementing a listener could become cumbersome.

davidolorundare commented:
hi @Wadeck , currently the only the `HudsonPrivateSecurityRealm` class makes use of the new methods. However, we foresee the methods being very useful for other authentication plugins that want to add support for notifying Jenkins of newly created accounts proactively, rather than waiting for someone to attempt to log in first.

Wadeck reviewed (dismissed):
Due to the merge of https://github.com/jenkinsci/jenkins/pull/3817, a new method in `HudsonPrivateSecurityRealm` was added to create user from a hashed password. Adding the event trigger in that new method is required to have this PR merged.

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare committed:
[JENKINS-55305] Add new method to SecurityListener (#3825)

Consists of three changes:

[JENKINS-55305] Update SecurityListener method JavaDoc

[JENKINS-55307] Update HudsonPrivateSecurityRealm method

[JENKINS-54965] Add new HudsonPrivateSecurityRealmTest unit-test

jeffret-b commented:
Latest updates look good.

davidolorundare commented:
cc @Wadeck updates have just been made to the PR

Wadeck commented:
@davidolorundare @jeffret-b 

> Fired after a new user account has been created and saved to disk.

I am sorry but this is not true. For your both usage of the even the user is created in memory at this point and not yet stored on disk.

Such description is especially useful for the listener implementation, and thinking the user is already saved but actually not really, could be dangerous.

Please either change the description or adapt the code. (The first one is safer)

jeffret-b commented:
After looking at it again, I'm not fond of what we changed the JavaDoc to say. I propose changing it to just say, "Fired after a new user account has been created." That seems more universally applicable and sufficiently defined.

If you've got different ideas @Wadeck, could you please clarify them? I'm not entirely sure what you're saying or suggesting.

If this seems like a reasonable approach, we can ask @davidolorundare to submit a minor PR to remove that phrase.

daniel-beck commented:
Is the problem here the uncertainty around the concept of 'user existence' when an external security realm is involved? I.e. a user may or may not exist at any time with external security realms (see e.g. legacy token auto creation).

Wadeck commented:
@daniel-beck that's exactly my point. The events must be defined more strictly. Look at the `loggedIn`, it specifies that:
> It should be called after the {@link org.acegisecurity.context.SecurityContextHolder#getContext()}'s authentication is set.
So, the consumer knows that in its implementation they will be able to request the authentication from the context. That's not the case for `authenticated`, but in practice that's mostly true.

As a event consumer, I am more than happy to see such things to avoid troubles. Without putting rules there, all the implementations **must** have multiple condition checks like does the user exist, do I need to create it, do I need to save it, can I use the session from the request, etc. And if one of the plugin is not firing the event at the correct location, the implementation will perhaps not work.

So, instead of reducing the JavaDoc, I prefer to have a correction there. In this case, we need to say that the user is created but not yet saved.

More generally for events, we need to ensure:
1) Most of the authentication plugin are able to fire that event (are they creating users ? do they save the user to disk themselves ? etc.)
2) The event listening is useful (=helpful) **and** usable (=can do thing with it) to achieve logic. I don't talk about just logging the username of the newly created account. Imagine if you want to add a property to the user, like the creation date, it will work. Now imagine you want to update the user mapping xml file, as the user is not saved, you need to be aware that this event will not fit your requirement.

Consequences of the two first points, due to the ecosystem:
3) We need to have multiple popular plugins that are using the new mechanisms
4) If there are popular plugins that are listening to those events, so the authentication plugins will fire the events. 

It's like a positive feedback cycle. Without plugins listening to the events, nobody will care.

daniel-beck commented:
> Most of the authentication plugin are able to fire that event

I'm not sure this would actually be useful. There's a huge difference between built in security realm notifications of user creation, and external ones.

Wadeck commented:
> [...] of user creation, and external ones.

When LDAP plugin received for the first time a user, it will create an account in the internal User database. From my understanding, this event should triggered in such case, no?

Otherwise the audit plugin will be able to follow the user creation only if you are using the embedded security realm, that is a bit restrictive IMHO.

daniel-beck commented:
> From my understanding, this event should triggered in such case, no?

I'm not sure this is relevant enough. A user might as well already exist in Jenkins from SCM based sources, so it's not even as if this would reliably indicate a 'first ever login'.

> Otherwise the audit plugin will be able to follow the user creation only if you are using the embedded security realm, that is a bit restrictive IMHO.

Users in external security realms just exist independent of Jenkins.

jvz commented:
External security realms have their own audit logging.

Wadeck commented:
As we can see with the discussion here, the JavaDoc must be rewritten. If it's specific for the embedded security realm, it must written explicitely.

jeffret-b commented:
> In this case, we need to say that the user is created but not yet saved.

Maybe I'm just not understanding what you precisely mean by "saved". In the specific case of `HudsonPrivateSecurityRealm` a file has been stored onto disk by the time `fireUserCreated()` is called. What additional conditions do you expect should exist before the user is considered to be "saved"?

(The flow and implementation of user handling has many structural problems. It would be nice to clean many parts of that up and make it sensible but that's difficult.)

We shouldn't expect that all providers have a distinct state of "saved" or that "created" is separate from it. With the `HudsonPrivateSecurityRealm` there isn't an easily recognized difference between those states. Not all services will persist data to disk to achieve persistence.

> 2\. Now imagine you want to update the user mapping xml file, as the user is not saved, you need to be aware that this event will not fit your requirement.

Any idea that this should be possible or meaningful is part of the structural flaws in handling users. Access to the operations should be via the service or API. Intimate, private knowledge should be discouraged, especially when it runs behind the interface.

> 1. Most of the authentication plugin are able to fire that event

This seems like an excessively high barrier to improvements. If we have to make sure that a significant percentage of plugins can and do participate, it makes it difficult to ever introduce anything new. We shouldn't introduce anything that doesn't have a use, but requiring that most existing ones participate locks us too heavily into our problematic legacy.

> do they save the user to disk themselves ?

How is that relevant to Jenkins? We should only be concerned about what is exposed through their interface and not how they implement and store things.

> From my understanding, this event should triggered in such case, no?

I don't think so. Maybe the JavaDoc can be improve to clarify that, but I can't quite figure out what it would say. There is already such a mess with account users, SCM users, and ephemeral users and unclear states and connections between them.


(I think you're actually addressing some different points, but I'm not understanding your concerns and intentions. Perhaps if you could propose what the JavaDoc should say or how the implementation should be changed that would help.)

Wadeck commented:
> > Fired after a new user account has been created and saved to disk.

> I am sorry but this is not true.

I put the breakpoint on the `addProperty` instead of putting on `fireUserCreated` method. So that confusion generated is totally my fault, sorry! The `addProperty` is triggering the `save` of the user to the disk.

So I am fine with that part of the JavaDoc.

===========

> > From my understanding, this event should triggered in such case, no?

> I'm not sure this is relevant enough

> External security realms have their own audit logging.

> I don't think so

As everyone seems to say that an external security realm should not be able to use the `fireUserCreated`, I think we can either restrict the access (`@Restricted(NoExternalUse.class)`) and mention in the JavaDoc, that only the embedded security realm is a producer of that event.

daniel-beck commented:
> an external security realm should not be able to use

A plugin could easily provide an alternative 'internal' user database with legitimate uses of `fireUserCreated`.

Wadeck commented:
OK let's stop the discussion there, let the JavaDoc as is and we will react in the future only if it's required.
